django/contrib/auth/__init__.py

```diff
@@ -91,6 +93,14 @@ def authenticate(request=None, **credentials):
     )
 
 
+@sensitive_variables("credentials")
+async def aauthenticate(request=None, **credentials):
+    """
+    If the given credentials are valid, return a User object.
+    """
+    return await sync_to_async(authenticate)(request, **credentials)
```

Comment:
It seems `@sensitive_variables` doesn't work with coroutines. For example, the following test crashes:
```diff
diff --git a/tests/auth_tests/test_auth_backends.py b/tests/auth_tests/test_auth_backends.py
index a535f329d2..81406f37e6 100644
--- a/tests/auth_tests/test_auth_backends.py
+++ b/tests/auth_tests/test_auth_backends.py
@@ -6,6 +6,7 @@ from django.contrib.auth import (
     BACKEND_SESSION_KEY,
     SESSION_KEY,
     _clean_credentials,
+    aauthenticate,
     authenticate,
     get_user,
     signals,
@@ -764,6 +765,28 @@ class AuthenticateTests(TestCase):
             status_code=500,
         )
 
+    @override_settings(
+        AUTHENTICATION_BACKENDS=["auth_tests.test_auth_backends.TypeErrorBackend"]
+    )
+    async def test_aauthenticate_sensitive_variables(self):
+        try:
+            await aauthenticate(
+                username="testusername", password=self.sensitive_password
+            )
+        except TypeError:
+            exc_info = sys.exc_info()
+        rf = RequestFactory()
+        response = technical_500_response(rf.get("/"), *exc_info)
+        self.assertNotContains(response, self.sensitive_password, status_code=500)
+        self.assertContains(response, "TypeErrorBackend", status_code=500)
+        self.assertContains(
+            response,
+            '<tr><td>credentials</td><td class="code">'
+            "<pre>&#39;********************&#39;</pre></td></tr>",
+            html=True,
+            status_code=500,
+        )
+
     def test_clean_credentials_sensitive_variables(self):
         try:
             # Passing in a list to cause an exception
```
We should start by fixing it in a separate PR :thinking:

Reply:
Oh wow! Nice catch. I've got a PR up that fixes this: https://github.com/django/django/pull/16668

Reply:
Hm but this still doesn't fix the test case you created. I'll see what's up there

Reply:
Alright I can't figure this out tonight, I'll take another look when I can

Reply:
Decorator is probably lost when we use `sync_to_async`:

![image](https://user-images.githubusercontent.com/2865885/226544128-b6b864c5-1628-46e6-b4af-5824e6e2ce1e.png)

Reply:
Oh then the problem isn't really the decorator handling coroutines (https://github.com/django/django/pull/16668) but more about how the implementation of `get_traceback_frame_variables` doesn't handle this case correctly? Am I on the right track here?

The reason I say this is I added some logging to `get_traceback_frame_variables` to print out the `current_frame` in the `while current_frame is not None` loop and it never actually hits the frames inside the asyncio executor (i.e. the one from your screenshot) which suggests to me that due to the way `asgiref` schedules the async function it might be missed by the traceback cleaning process. (I'm a little out of my depth here, I may be saying nonsense)

Reply:
As far as I'm aware, `sync_to_async` call should also be wrapped with `sensitive_variables` somehow :thinking: as I found this frame in the traceback:
```
<frame at 0x7f7e4e187840, file '/python3.10/site-packages/asgiref/sync.py', line 495, code thread_handler>

dict_items([
('self', <asgiref.sync.SyncToAsync object at 0x7f7e4e1a77f0>),
('loop', <_UnixSelectorEventLoop running=True closed=False debug=False>),
('source_task', <Task pending name='Task-1' coro=<AsyncToSync.main_wrap() running at /python3.10/site-packages/asgiref/sync.py:306> cb=[_run_until_complete_cb() at /python3.10/asyncio/base_events.py:184]>),
('exc_info', (None, None, None)),
('func', <built-in method run of _contextvars.Context object at 0x7f7e4e2a6a40>),
('args', (functools.partial(<function authenticate at 0x7f7e4f29e4d0>, None, username='testusername', password='mypassword'),)),
('kwargs', {}),
('current_thread', <_MainThread(MainThread, started 140180512552768)>),
('parent_set', False)])
```
and the previous frame is
```
<frame at 0x7fa4e89be3e0, file '/python3.10/site-packages/asgiref/current_thread_executor.py', line 26, code run>
```
instead of
```
<frame at 0x7fa4e899ba00, file '/django/django/views/decorators/debug.py', line 56, code sensitive_variables_wrapper>
```

Reply:
Grrr. I just merged in #16668 having played with it this morning — the behaviour is OK there with the explicit `async def` declaration... — We may have to adjust that. 🤔

> Oh then the problem isn't really the decorator handling coroutines (https://github.com/django/django/pull/16668) but more about how the implementation of get_traceback_frame_variables doesn't handle this case correctly? Am I on the right track here?

Yes, there's something going on there. Musing aloud... once you hit the suspension at an `await` the stack is _"rebuilt"_ for the task to execute before the continuation puts you back where you were. It's not 100% clear to me that we can dig around in private parts of the traceback and expect to find exactly what we're looking for. (Need to dig more.)

Reply:
Grrr. For four or five frames going through SyncToAsync the passed parameters are exposed, like... 

```
args | (functools.partial(<function authenticate at 0x106eecd60>, None, username='testusername', password='mypassword'),)
```

and similar. 

Once the other side they are handled again in the lower frames. 

There's no obvious way around this. A exception report filter that kept a track of sensitive ways, and knew to deal with SyncToAsync... — bah! More likely a fully async authentication backend.

Reply:
> Grrr. I just merged in https://github.com/django/django/pull/16668 having played with it this morning — the behaviour is OK there with the explicit async def declaration... — We may have to adjust that. thinking

I think it does need adjustment, but not

Reply:
Ok I think I have *a* solution. Not a great one, but it does actually work without requiring asyncification of all the `auth` internals at once!

I borrowed this technique from my day job: avoid assigning or logging a sensitive variable by just storing the sensitive variable in the return value of a lambda. (see below for this concept applied here).

This works just fine, but unfortunately the error reporter still isn't tuned correctly:

> Grrr. I just merged in https://github.com/django/django/pull/16668 having played with it this morning — the behaviour is OK there with the explicit async def declaration... — We may have to adjust that. thinking

Yeah unfortunately that looks like it might have been a partial solution, though the missing piece evades me. The full solution won't fix the `asgiref.sync` helpers in any event so I think  https://github.com/django/django/commit/b00046d2c25771bed2242680b08b524a44aa9798 is a good move, but there is a situation that my test case didn't cover that I can't quite pin down.

I took @felixxm's suggested test case and merged it to my current commit, then I added a new commit which just contains the "variable stored in a lambda" technique that I talked above above. I reran the test case Mariusz wrote and that commit reduced the rate of "mypassword" appearing from several (the `asgiref` stack) down to a single hit. This doesn't make sense to me as it should be cleansed by `@sensitive_variables` but it isn't.

When I was digging in to it I zeroed in on how `technical_500.html` tracebacks are actually generated and I noticed that, well first I have no idea how any of this code works, but after some very dumb debug `print`ing I noticed that `aauthenticate` doesn't have an `f_back` in the `tb_frame` object (...I started regretting opening this PR at this point). On a lark I passed in the previous `tb` and massaged it in to working. So I committed that fix to the third commit in the stack. I don't think this is the right solution (and not backwards compatible), but I'm not really sure why this "fix" is even necessary in the first place.

I'll have to take another look with fresh eyes sometime later (likely next week at the earliest), but I wanted to get this back up for another round of discussions, and maybe someone else has an idea on why commit #3 is necessary.

In summary, I have pushed the following commits:
1. The actual changes for this PR (including Mariusz's test case, which should stay no matter what)
2. The "variable stored in a lambda" magic, which I'm ok with but might be a bit too far of a reach
3. A maybe sorta hacky fix to the magical traceback machinery

IMO (2) should merge in to (1), if we agree that this is the right path, and (3) should be... rewritten.

Reply:
I've made the tiniest amount of progress. The following patch can be applied against prod and causes the `view_tests.tests.test_debug.ExceptionReporterFilterTests.test_async_sensitive_request` test I case that I wrote in https://github.com/django/django/pull/16668 to fail. I'm not sure how exactly to fix this though, but I'm staring at it very hard. if I can't figure this out we should probably revert that PR since it only protects the named sensitive variables if an exception is thrown *by the function wrapped by `@sensitive_variables`* which is basically no protection at all. 

```diff
diff --git a/tests/view_tests/views.py b/tests/view_tests/views.py
index 97febdaf83..1f01e31504 100644
--- a/tests/view_tests/views.py
+++ b/tests/view_tests/views.py
@@ -180,7 +180,7 @@ def sensitive_view(request):
 
 @sensitive_variables("sauce")
 @sensitive_post_parameters("bacon-key", "sausage-key")
-async def async_sensitive_view(request):
+async def async_nested_view(request):
     # Do not just use plain strings for the variables' values in the code
     # so that the tests don't return false positives when the function's source
     # is displayed in the exception report.
@@ -188,8 +188,13 @@ async def async_sensitive_view(request):
     sauce = "".join(  # NOQA
         ["w", "o", "r", "c", "e", "s", "t", "e", "r", "s", "h", "i", "r", "e"]
     )
+
+    raise Exception
+
+
+async def async_sensitive_view(request):
     try:
-        raise Exception
+        return await async_nested_view(request)
     except Exception:
         exc_info = sys.exc_info()
         send_log(request, exc_info)
```

Reply:
> If I can't figure this out we should probably revert that PR since it only protects the named sensitive variables if an exception is caught by the function wrapped by `@sensitive_variables` which is functionally not enough protection to warrant documentation that we DO support this.

Agreed, any error-prone solution in this area can be really damaging.

Reply:
See #16704.

Reply:
Ok this issue has been resolved now due to the changes in https://github.com/django/django/commit/38e391e95fe5258bc6d2467332dc9cd44ce6ba52